README.md
# Overview
This repository contains everything necessary to convert Corelight or Zeek Logs into the Elastic Common Schema (ECS) naming standard and store them into an Elastic Stack deployment. Originally released in 2019, the same year as Elastic Common Schema, this repository is stil actively maintained and updated.

It supports over [200 different logs and metrics](https://docs.google.com/spreadsheets/d/1HqRRfvf5vUG3zgI9KG64hAj5oLSae7YbLoTKh846FCc/edit?usp=sharing) and is designed to be able to support all environments while simulatenously being able to customize the required and desired options to support each unique environment.  
Significant effort has been invested in ensuring support across all environments, while simultaneously allowing for the customization of necessary and desired options to suit each unique and respective deployment. 


The repository is primiarly compromised of two things:
1. Elasticsearch index templates, component templates, ilm policies, settings, and mappings
2. Installer script for this repository and the corresponding [Ingest Pipelines](https://github.com/brasitech/ecs-mapping) OR [Logstash Pipelines](https://github.com/brasitech/ecs-logstash-mappings)

# License
The files and automation script are open-source under a BSD license. See ``COPYING``for details.

# Install Steps
**Please read all sections before running the script.**
1. [Requirements](#1-requirements)
2. [Customization](#2-customization)
3. [Install](#3-install)
4. [Post Install](#4-post-install)

 
# 1. Requirements
- Using the [`corelight_ecs.py`](./corelight_ecs.py) script from this repository. It is not supported to manually install the templates and pipelines because there hundreds of variables throughout the files that get replaced by the script.
- Python 3.6 and higher
- Elasticsearch version 8.x and higher or if still on version 7.x than 7.17 and higher
- If using Logstash, than Logstash version 8.x and higher or if still on version 7.x than 7.17 and higher
- Elasticsearch Datastream indexing strategy


# 2. Customization
Customizing Index Settings, Mappings, Aliases, and ILM Policies is encouraged. We all know how necessary it is. Therefore, the script is designed to be able to support the following but not limited to, types of customizations.
- [Custom Index Names](#custom-index-names)  
  the script will prompt you for this and automatically update any templates, logstash pipelines, and or ingest pipelines that need to be updated
- [Index Settings](#custom-index-settings-ie-shards-and-replicas)
  - Number of shards
  - Number of replicas
  - Override other index settings
  - Anything else you see fit in the component template for the settings portion
- [Index Mappings](#customizing-index-mappings-ie-fields)
    - Override Elasticsearch index field mapping
    - Anything else you can see fit in the component template for the mappings portion
- [ILM Policies](#custom-ilm-policy-strategy)
  - Custom ILM policy
- [Index Aliases](#custom-index-aliases)
  - Alaises for the index patterns, such as for backwards compatibility or to disable it
- [Custom Ingest Pipelines](#custom-ingest-pipelines)
    - The ability to call your own custom ingest pipeline before any Corelight ingest pipelines are called and without modifying the Corelight ingest pipelines at all
    - The ability to call your own custom ingest pipeline after all Corelight ingest pipelines are called and without modifying the Corelight ingest pipelines at all
- Custom logstash pipelines is just matter of creating a file with a lower alphanumeric order and loading in the same directory

Many customizations will not be needed, but it is better to have the option than have not. Always recommend to use your own number of shards/replicas and ILM policy, but beyond that you should not need to customize much else you normally customize many things in the index templates, ingest pipelines, and logstash pipelines in all of your deployments. Which, in that case you will have that ability as well.

See each section below for more details.

## Custom Index Options

### Custom Index Names
The script will prompt you if you would look to choose custom index names/patterns. This, allows you to choose your own index naming convention. For example you could chose:
- `logs-corelight.conn-default`
- `ecs-corelight.conn.default`
- `call_it_what_you_will.conn.default`

It will then update all the necessary index templates as well if you chose ingest pipelines or logstash pipelines.
It's not recommended to change the defaults, as the Corelight logs are now setup to work with all the other types of ECS data, and be able to be used from within a Kibana data view such as `logs-*`

For multi-tenant environments such as MSSPs/CSSPs or different networks, you can not only choose the organization portion of the data stream in the script prompts (which is the `default` portion in the 3 examples above), however if you set data_stream.namespace in filebeat or some other method or hook the pipelines, then the pipelines are set to respect if that variable is already set to not override it. Just make sure that your namespace is set to a value that Elasticsearch allows for an index name (common mistake that is not obvious sometimes).

You can see the index patterns that get used in the [Google Sheet](https://docs.google.com/spreadsheets/d/1HqRRfvf5vUG3zgI9KG64hAj5oLSae7YbLoTKh846FCc/edit?usp=sharing) and the default for each log. Also, lists some additional information such as event.category, event.kind, and etc for each log.

### Custom Index Settings ie: Shards and Replicas
Each index template is specified to call a component template that you can use to override the Index Settings.
**Therefore you do not need to modify the repository index templates in order to use your own Index Settings.**

Create the following component templates or modify them post install. The name of the component templates that you can modify are:

_to cover all logs (split in 5 categories):_
- `corelight-ecs-component-metric_log-base-settings@custom`
- `corelight-ecs-component-parse_failures-base-settings@custom`
- `corelight-ecs-component-protocol_log-base-settings@custom`
- `corelight-ecs-component-system_log-base-settings@custom`
- `corelight-ecs-component-unknown_log-base-settings@custom`  

_to cover individual logs beyond the above, that commonly need their own settings to be even more specific:_
- `corelight-ecs-component-protocol_log-conn-base-settings@custom`
- `corelight-ecs-component-protocol_log-dns-base-settings@custom`
- `corelight-ecs-component-protocol_log-files-base-settings@custom`
- `corelight-ecs-component-protocol_log-http-base-settings@custom`
- `corelight-ecs-component-protocol_log-smb-base-settings@custom`
- `corelight-ecs-component-protocol_log-smtp-base-settings@custom`
- `corelight-ecs-component-protocol_log-ssl-base-settings@custom`
- `corelight-ecs-component-protocol_log-suricata_corelight-base-settings@custom`
- `corelight-ecs-component-protocol_log-syslog-base-settings@custom`
- `corelight-ecs-component-protocol_log-various-base-settings@custom`
- `corelight-ecs-component-protocol_log-weird-base-settings@custom`
- `corelight-ecs-component-protocol_log-x509-base-settings@custom`
- 
### Customizing Index Mappings ie: fields
Each index template is specified to call a component template that you can use to override the Index Mappings for things such as field types or names.
**Therefore you do not need to modify the repository index templates in order to use your own Index Mappings.**

Create the following component templates or modify them post install. The name of the component templates that you can modify are:

_to cover all logs (split in 5 categories):_
- `corelight-ecs-component-metric_log-mappings@custom`
- `corelight-ecs-component-parse_failures-mappings@custom`
- `corelight-ecs-component-protocol_log-mappings@custom`
- `corelight-ecs-component-system_log-mappings@custom`
- `corelight-ecs-component-unknown_log-mappings@custom`

_to cover individual logs beyond the above, that commonly need their own settings to be even more specific:_
- `corelight-ecs-component-protocol_log-conn-mappings@custom`
- `corelight-ecs-component-protocol_log-dns-mappings@custom`
- `corelight-ecs-component-protocol_log-files-mappings@custom`
- `corelight-ecs-component-protocol_log-http-mappings@custom`
- `corelight-ecs-component-protocol_log-smb-mappings@custom`
- `corelight-ecs-component-protocol_log-smtp-mappings@custom`
- `corelight-ecs-component-protocol_log-ssl-mappings@custom`
- `corelight-ecs-component-protocol_log-suricata_corelight-mappings@custom`
- `corelight-ecs-component-protocol_log-syslog-mappings@custom`
- `corelight-ecs-component-protocol_log-various-mappings@custom`
- `corelight-ecs-component-protocol_log-weird-mappings@custom`
- `corelight-ecs-component-protocol_log-x509-mappings@custom`


### Custom ILM Policy Strategy
Each index template is specified to call a component template that you can use to override the ILM Policy to set your own retention.
**Therefore you do not need to modify the repository index templates in order to use your own ILM Policy.**

Create the following component templates or modify them post install. The name of the component templates that you can modify are:

_to cover all logs (split in 5 categories):_
- `corelight-ecs-component-metric_log-ilm-settings@custom`
- `corelight-ecs-component-parse_failures-ilm-settings@custom`
- `corelight-ecs-component-protocol_log-ilm-settings@custom`
- `corelight-ecs-component-system_log-ilm-settings@custom`
- `corelight-ecs-component-unknown_log-ilm-settings@custom`

these will cover individual logs beyond the above, that commonly need their own settings to be even more specific:
- `corelight-ecs-component-protocol_log-conn-ilm-settings@custom`
- `corelight-ecs-component-protocol_log-dns-ilm-settings@custom`
- `corelight-ecs-component-protocol_log-files-ilm-settings@custom`
- `corelight-ecs-component-protocol_log-http-ilm-settings@custom`
- `corelight-ecs-component-protocol_log-smb-ilm-settings@custom`
- `corelight-ecs-component-protocol_log-smtp-ilm-settings@custom`
- `corelight-ecs-component-protocol_log-ssl-ilm-settings@custom`
- `corelight-ecs-component-protocol_log-suricata_corelight-ilm-settings@custom`
- `corelight-ecs-component-protocol_log-syslog-ilm-settings@custom`
- `corelight-ecs-component-protocol_log-various-ilm-settings@custom`
- `corelight-ecs-component-protocol_log-weird-ilm-settings@custom`
- `corelight-ecs-component-protocol_log-x509-ilm-settings@custom`



### Custom Index Aliases
Each index template is specified to call a component template that you can use to override the Index Alaises.
**Therefore you do not need to modify the repository index templates in order to use your own ILM Alaises.**

Create the following component templates or modify them post install. The name of the component templates that you can modify are:

_to cover all logs (split in 5 categories):_
- `corelight-ecs-component-metric_log-aliases@custom`
- `corelight-ecs-component-parse_failures-aliases@custom`
- `corelight-ecs-component-protocol_log-aliases@custom`
- `corelight-ecs-component-system_log-aliases@custom`
- `corelight-ecs-component-unknown_log-aliases@custom`

_these will cover individual logs beyond the above, that commonly need their own settings to be even more specific:_
- `corelight-ecs-component-protocol_log-conn-aliases@custom`
- `corelight-ecs-component-protocol_log-dns-aliases@custom`
- `corelight-ecs-component-protocol_log-files-aliases@custom`
- `corelight-ecs-component-protocol_log-http-aliases@custom`
- `corelight-ecs-component-protocol_log-smb-aliases@custom`
- `corelight-ecs-component-protocol_log-smtp-aliases@custom`
- `corelight-ecs-component-protocol_log-ssl-aliases@custom`
- `corelight-ecs-component-protocol_log-suricata_corelight-aliases@custom`
- `corelight-ecs-component-protocol_log-syslog-aliases@custom`
- `corelight-ecs-component-protocol_log-various-aliases@custom`
- `corelight-ecs-component-protocol_log-weird-aliases@custom`
- `corelight-ecs-component-protocol_log-x509-aliases@custom`

## Custom Ingest Pipelines

### Custom Beginning Ingest Pipeline Catchall
At the **beginning** of the ingest pipelines, before any Corelight pipelines are called, there is a function to use the ingest pipeline named `corelight-ecs-main-pipeline@custom`.
You can create an ingest pipeline with that name if there are things you want to do before the Corelight pipelines are called.
If you do not create the pipeline the function silently passes it, therefore it is optional.

### Custom Final Ingest Pipeline Catchall
At the **end** of the ingest pipelines, after all Corelight pipelines are called, there is a function to use the ingest pipeline named `corelight-ecs-postprocess-final-main-pipeline@custom`.
You can create an ingest pipeline with that name if there are things you want to do before the Corelight pipelines are called.
If you do not create the pipeline the function silently passes it, therefore it is optional.

### Custom Beginning Ingest Pipeline Parse Failure Catchall
At the **beginning** of the ingest pipelines on_failure catchall for ingest pipeline failures, before any Corelight pipelines are called, there is a function to use the ingest pipeline named `corelight-ecs-parse_failures-pipeline@custom`.
You can create an ingest pipeline with that name if there are things you want to do before the Corelight pipelines are called.
If you do not create the pipeline the function silently passes it, therefore it is optional.

### Custom Final Ingest Pipeline Parse Failure Catchall
At the **end** of the ingest pipelines, after all Corelight pipelines are called, there is a function to use the ingest pipeline named `corelight-ecs-postprocess-final-main-pipeline@custom`.
You can create an ingest pipeline with that name if there are things you want to do before the Corelight pipelines are called.
If you do not create the pipeline the function silently passes it, therefore it is optional.



changelog.md
# Version 20231201
- Minimum requirements for all software (elasticsearch, logstash, kibana) is 7.17

    - for many reasons, but 1 of many great reasons is Ingest pipeline => 7.11 supports the set processor function "copy\_from"

    - tag\_with\_exception\_message for ruby filter block

        - requires version => 7.17
        - requires version => 8.x

    - ingest pipelines finally performant in => 8.8

    - logstash more performant in 8.x

    - many many other reasons

- \[@metadata]\[corelight\_env\_vars]\[disable\_legacy\_md5\_fingerprint\_hash] if set to true will disable legacy md5 fingerprint hashes else will call 8811-corelight-ecs-network\_fingerprints-enrich-legacy\_md5-filter.conf and add an array w/ md5 to the existing sha1

- changed values for `labels.corelight.ecs_method`. values are now `logstash_pipeline` or `ingest_pipeline` previously were `logstash` or `ingest_pipeline`
- added `labels.corelight.ecs_pipeline_method`
- replaced text analyzer with es\_stk\_analyzer
- removed stats index, replaced with metrics. event.category is diagnostics same as the other metrics (previously was 'stats') these were the logs:
    - capture_loss
    - corelight_cloud_stats
    - corelight_overall_capture_loss
    - corelight_profiling
    - corelight_weird_stats
    - namecache
    - packet_filter
    - reporter
    - stats
    - suricata_stats
    - weird_stats
- removed netcontrol and iam index, replaced with system. these were the logs:
    - netcontrol
    - netcontrol_drop
    - netcontrol_shunt
    - corelight_audit_log
    - audit
    - auditlog
- changed suricata_corelight dataset back to suricata_corelight from suricata
- parse-failures is now parse\_failures & parse-failure is parse\_failure

- change intel log event.category from intrustion_detection to threat

- changed VAR\_$Corelight\_LS\_Index\_Strategy to VAR\_CORELIGHT\_INDEX\_STRATEGY

- logs that do not explicitly have a pipeline/parser (new or unknown) they get set to event.category tbd (previously used either temporary or event.dataset). not use event.dataset as do not want to autocreate datastreams that would live outside ILM patterns and other control/settings and permissions. can still search the logs as is despite the name in the index pattern. event.dataset still gets set.

- you can set \[@metadata]\[custom\_temporary\_metadata\_index\_name\_namespace] or \[data\_stream]\[namespace] to set a desired namespace for multiple tenants/clients data

    - if \[@metadata]\[custom\_temporary\_metadata\_index\_name\_namespace] is set then it overrides if \[data\_stream]\[namespace] already exists

- for custom mappings,settings,ilm,etc.. use

    - aliases

        - corelight-main\_logs-aliases\@custom
        - corelight-metrics\_and\_stats-aliases\@custom
        - corelight-parse\_failures-aliases\@custom

    - base settings

        - corelight-main\_logs-base-settings\@custom
        - corelight-metrics\_and\_stats-base-settings\@custom
        - corelight-parse\_failures-base-settings\@custom

    - ilm settings

        - corelight-main\_logs-ilm-settings\@custom
        - corelight-metrics\_and\_stats-ilm-settings\@custom
        - corelight-parse\_failures-ilm-settings\@custom

    - mappings

        - corelight-main\_logs-mappings\@custom
        - corelight-metrics\_and\_stats-mappings\@custom
        - corelight-parse\_failures-mappings\@custom

- ingest pipelines custom additions user controlled\
  if the pipelines do not exist, it just ignores them. so in other words, is optional for user

- names are

    - before starting corelight ingest pipelines use 'corelight-ecs-main-pipeline@custom'
    - after completing corelight ingest pipelines use 'corelight-ecs-postprocess-final-main-pipeline@custom'

- date fields that are user/software/attacker controlled (smtp.date for example) or commonly could be invalid (like for x509 cert dates before/after epoch) are always kept now instead of removed. if valid date timestamp then it is copied as such to a new field

- logstash moved 0101-corelight-ecs-user\_defined-set\_indexing\_strategy-filter.conf into 3100-corelight-ecs-common-set\_index\_prefix\_and\_suffix-filter.conf ( no longer need to edit manually anyways)

- installer script, use last run

    - uses last run, allows directory to be specified
    - this can also be used to just use the script to upload things you generated elsewhere or if already have the files generated

- additional changelog

    - logstash network fingerprints have md5 and sha1, sha1 will be used for future so that ingest pipelines will be able to support to. and ingest pipelines do not have a builtin callable md5 from script processor. only fingerprint processor that unfortunately has a bug of copying a null byte that makes the hash unusuable with other things (like logstash in this case)

    - single quote invalid IPs in event.tag (previously was not quoted)

    - IPs beginning with "0" (ie: 0.0.0.0/8) changed

        - ip\_type from "reserved\_as\_a\_source\_address\_only" to "reserved\_local\_this\_network"
        - ip\_rfc from "RFC\_1700" to "RFC\_1122-3.2.1.3"

    - split out RFC info for all RFC6890, RFC2544, RFC5737 that were previously tagged together

    - 255.255.255.255 changed from RFC\_8190 to RFC\_919

    - added rfc for 240.0.0.0/4 RFC5735, no longer grouped under multicast

    - carrier grade nat 100.64.0.0-100.127.255.255 fixed type label

    - 0.0.0.0 set to type reserved\_any\_address

    - switched ingest pipeline's main\_pipeline to call pipeline name w/ the event.dataset

    - changed the "genera"/"common" corelight\_genenral\_pipeline to corelight\_common\_pipeline

    - renamed all pipelines (to match similar naming in logstash)

        - corelight\_ to corelight-ecs-
        - \_pipeline to -pipeline

    - renamed pipelines (to match similar naming in logstash)

        - metrics\_general to common-metrics
        - netcontrol\_general to common-netcontrol

installer.py
#!/usr/bin/env python3

#from asyncore import file_wrapper
#from distutils import filelist
#from this import d, s
#import glob
import shutil
import requests
from urllib3.exceptions import InsecureRequestWarning
import time
import sys
import zipfile
import os
import random
import subprocess

def checkRequest(responseObj):
    code = responseObj.status_code
    if code == 200:
        return 200

    if 400 <= code <= 500:
        print(responseObj.json())
        time.sleep(5)
        return code
    return code

def input_bool(question, default=None):

    prompt = " [yn]"

    if default is not None:
        prompt = " [Yn]:" if default else " [yN]:"

    while True:
        val = input(question + prompt)
        val = val.lower()
        if val  == '' and default is not None:
            return default
        if val in ('y', 'n'):
            return val == 'y'
        print("Invalid response")

def input_int(question):
    while True:
        val = input(question + ": ")
        try:
            return int(val)
        except ValueError as e:
            print("Invalid response", e)

def testConnection(session, baseURI):
    testUri = "/_cat/indices?v&pretty"
    uri = baseURI + testUri
    response = session.get(uri, timeout=5)
    checkRequest(response)
    response.raise_for_status()

def download_repository( name ):
    randomNum = random.randint(0,100)
    filename = "output" + str(randomNum) + ".zip"
    host = "https://github.com/"
    organization = "corelight/"
    url = host + organization + name 
    with requests.get(url, stream=True) as r:
        with open(filename, 'wb') as f:
            shutil.copyfileobj(r.raw, f)
    return filename

def unzipGit(filename):
    with zipfile.ZipFile(filename, 'r') as zip_ref:
        zip_ref.extractall()

def installLogstash(directory):
    source = "./ecs-logstash-mappings-Dev/pipeline/"
    if directory.endswith('/'):
        path = directory + "CorelightPipelines"
    else:
        path = directory + "/CorelightPipelines"
    if os.path.exists(directory):
        if not os.path.exists(path):
            shutil.copytree(source, path)
        else:
            print("the path %s exists have if you are updating the please select the update operation" % path)
            sys.exit(1)
    else:
        print("The path you entered does not exist")
        sys.exit(1)    
    
def updateLogstash(directory):
    source = "./ecs-logstash-mappings-Dev/pipeline/"
    tcp = ""
    path = directory + "/CorelightPipelines"
    if os.path.exists(directory):
        fileList=os.listdir(source)
        for filename in fileList:
            shutil.copy2(os.path.join(source,filename), path)

def enableIngest(ingest_type, raw, logstashLocation):
    ls_pipeline_install_dir = logstashLocation + "/CorelightPipelines/"
    tcp = "0002-corelight-ecs-tcp-input.conf"
    ssl = "0002-corelight-ecs-tcp-ssl_tls-input.conf"
    hec = "0002-corelight-ecs-http-for_splunk_hec.conf"
    kafka = "0002-corelight-ecs-kafka-input.config"
    tcpRaw = "0002-corelight-ecs-tcp-input-codec_disabled_to_keep_raw_message.conf"
    sslRaw = "0002-corelight-ecs-tcp-ssl_tls-input-codec_disabled_to_keep_raw_message.conf"
    hecRaw = "0002-corelight-ecs-http-for_splunk_hec-codec_disabled_to_keep_raw_message.conf"
    kafkaRaw = "0002-corelight-ecs-kafka-input-codec_disabled_to_keep_raw_message.config"
    
    if raw:
        if ingest_type == "tcp":
            source = ls_pipeline_install_dir + tcpRaw + ".disabled"
            dest = ls_pipeline_install_dir + tcpRaw 
        if ingest_type == "ssl":
            source = ls_pipeline_install_dir + sslRaw + ".disabled"
            dest = ls_pipeline_install_dir + sslRaw 
        if ingest_type == "hec":
            source = ls_pipeline_install_dir + hecRaw + ".disabled"
            dest = ls_pipeline_install_dir + hecRaw 
        if ingest_type == "kafka":
            source = ls_pipeline_install_dir + kafkaRaw + ".disabled"
            dest = ls_pipeline_install_dir + kafkaRaw 
    else:
        if ingest_type == "tcp":
            source = ls_pipeline_install_dir + tcp + ".disabled"
            dest = ls_pipeline_install_dir + tcp
        if ingest_type == "ssl":
            source = ls_pipeline_install_dir + ssl + ".disabled"
            dest = ls_pipeline_install_dir + ssl
        if ingest_type == "hec":
            source = ls_pipeline_install_dir + hec + ".disabled"
            dest = ls_pipeline_install_dir + hec
        if ingest_type == "kafka":
            source = ls_pipeline_install_dir + kafka + ".disabled"
            dest = ls_pipeline_install_dir + kafka 
    shutil.copy(source, dest)

def postPorcessing(logstashLocation, datastream, logstashVersion):
    ls_pipeline_install_dir = logstashLocation + "/CorelightPipelines/"
    if logstashVersion:
        sedCommand = 'sed -i "s/#ecs_compatibility =>/ecs_compatibility =>/" ' + ls_pipeline_install_dir + '/*.conf'
        subprocess.call([sedCommand],shell=True)
    if datastream:
        filename = ls_pipeline_install_dir + "0101-corelight-ecs-user_defined-set_indexing_strategy-filter.conf.disabled"
        f = open(filename)
        ds = f.read()
        if '=> "$Corelight_LS_Index_Strategy"' in ds:
            dsEnabled = ds.replace('=> "$Corelight_LS_Index_Strategy"', '=> "datastream"')
            f.close()
            dsOut = ls_pipeline_install_dir + "0101-corelight-ecs-user_defined-set_indexing_strategy-filter.conf"
            f = open(dsOut, "wt")
            f.write(dsEnabled)
            f.close()

def exportToElastic(session, baseURI, filePath, pipeline, path,  retry=4):
    filename = filePath + pipeline
    if pipeline != "zeek-enrichment-conn-policy/_execute":
        with open(filename) as f:
            postData = f.read()
    else:
        postData = ""
    run = 1
    uri = baseURI + path + pipeline
    
    response = 0
    while run <= retry and response != 200:
        run = run + 1
        response = session.put(uri, data=postData, timeout=10)
        response = checkRequest(response)
        if response == 400 or response == 409:
            print(response)

    if response == 200:
        return 
    else:
        print("Error uploading %s status code %s" %(pipeline, response),file=sys.stderr)
        print("URI = %s" % uri)
        sys.exit(1)


def elasticDel(session, baseURI, pipeline,  retry):

    uri = baseURI + "/_ingest/pipeline/"  + pipeline
    if pipeline.endswith("-policy"):
        uri = baseURI + "/_enrich/policy/" + pipeline

    print("deleting uri = %s" % uri)
    response = session.delete(uri, timeout=5)
    result = checkRequest(response)
    return result

def get_config():
    """Return a baseURI and session"""

    s = requests.Session()
    s.headers={'Content-Type': 'application/json'}

    print("\nEnter the information to connect to your Elasticsearch cluster.\n")
    ipHost = input("Hostname or IP: ")
    port = input_int("Port")
    auth = input_bool("Use user and password authentication?", default=True)

    if auth:
        user = input("User: ")
        password = input("Password: ")
        s.auth = (user, password)
    secure = input_bool("Use https?")
    ignoreCertErrors = False
    if secure:
        ignoreCertErrors = input_bool("Would you like to ignore certificate errors?", default=False)
    s.verify = not ignoreCertErrors

    print("You have entered the following parameters to connect to your cluster: \n - Host/IP: %s \n - Port: %s \n - HTTP/HTTPS: %s" %(ipHost, port, secure))
    if auth:
        print(" -User: %s \n -Password: %s" %(user, password))
    if secure:
        print(" - Ignore Certificate Errors: %s" % ignoreCertErrors)

    proto = "https" if secure else "http"
    baseURI = proto + "://" + ipHost + ":" + str(port)
    return baseURI, s

def datastreams(session, baseURI, logstash,updateTemplates):
    source = "./templates-component/data_stream/"
    component = source + "component_template/"
    ilm = source + "ilm_policy/"
    index = source + "index_template/"
    ingest = source + "use_ingest_pipeline/"
    fileList=os.listdir(component)
    for filename in fileList:
        exportToElastic(session, baseURI, component, filename, "/_component_template/", retry=4)
    if not updateTemplates:
        fileList=os.listdir(ilm)
        for filename in fileList:
            exportToElastic(session, baseURI, ilm, filename, "/_ilm/policy/", retry=4)
    
    fileList=os.listdir(index)
    for filename in fileList:
        exportToElastic(session, baseURI, index, filename, "/_index_template/", retry=4)
    if not logstash:
        exportToElastic(session, baseURI, ingest, "corelight-ds-component_template-use_ingest_pipeline-settings", "/_component_template/", retry=4)
        # exportToElastic(session, baseURI, ingest, "corelight-ds-index_template-main_logs_use_ingest_pipeline", "/_index_template/", retry=4)
        exportToElastic(session, baseURI, ingest, "corelight-ds-index_template-metrics_and_stats", "/_index_template/", retry=4)
        exportToElastic(session, baseURI, ingest, "corelight_postprocess_index_naming_pipeline", "/_ingest/pipeline/", retry=4)

def component( session, baseURI, logstash, updateTemplates ):
    source = "./templates-component/non_data_stream/"
    component = source + "component_template/"
    ilm = source + "ilm_policy/"
    index = source + "index_template/"
    ingest = source + "use_ingest_pipeline/"
    fileList=os.listdir(component)
    for filename in fileList:
         exportToElastic(session, baseURI, component,filename, "/_component_template/", retry=4)
    if not updateTemplates:
        fileList=os.listdir(ilm)
        for filename in fileList:
            exportToElastic(session, baseURI, ilm, filename, "/_ilm/policy/", retry=4)
    fileList=os.listdir(index)
    for filename in fileList:
        exportToElastic(session, baseURI, index, filename, "/_index_template/", retry=4)
    if not logstash:
        exportToElastic(session, baseURI, ingest, "corelight-non-ds-component_template-use_ingest_pipeline-settings", "/_component_template/", retry=4)
        exportToElastic(session, baseURI, ingest, "corelight-non-ds-index_template-main_logs_use_ingest_pipeline", "/_index_template/", retry=4)
        exportToElastic(session, baseURI, ingest, "corelight-non-ds-index_template-metrics_and_stats", "/_index_template/", retry=4)

def index(session, baseURI, logstash,updateTemplates):
    source = "./templates-component/templates-legacy/"
    ingest = source + "use_ingest_pipeline/"
    fileList=os.listdir(source)
    for filename in fileList:
        exportToElastic(session, baseURI, source, filename, "/_template/", retry=4)
    
    if not logstash:
        fileList=os.listdir(ingest)
        for filename in fileList:
                exportToElastic(session, baseURI, ingest, filename, "/_template/", retry=4)

def uploadIngestPipelines(session,baseURI):
    source = "./ecs-mapping-master/automatic_install/"
    fileList=os.listdir(source)
    for filename in fileList:
        if "deprecated" not in filename:
            exportToElastic(session, baseURI, source, filename, "/_ingest/pipeline/", retry=4)


def main():

    updateTemplates = False
    logstashRepo="ecs-logstash-mappings/archive/refs/heads/Dev.zip"
    ingestRepo="ecs-mapping/archive/refs/heads/master.zip"
    requests.packages.urllib3.disable_warnings(category=InsecureRequestWarning)
    baseURI, session = get_config()
    testConnection(session, baseURI)
    logstash = input_bool("Will you be using a Logstash pipeline?", default=True)
    templatesOnly = input_bool("Will you only be installing Templates, if using ingest Pipeline enter No?", default=False)
    if templatesOnly:
        updateTemplates = input_bool("Is this a update to existing template? If so ILM will not be installed", default=False)
    if not updateTemplates:
        if logstash:
            cont = input_bool("This script needs to be run on the Logstash box. Does this box have Logstash running and is the Logstash ingesting?", default=True)
            if cont:
                #Answer Yes if you are in a offline or Airgaped enviorment 
                # Then give the filename and location and it will run
                download = input_bool("Have you downloaded logstash repo?", default=False)
                if download:
                    filename = input("Please enter the filename for the Zip file of the Logstash Pipeline repo? <note it need to be in the same directorey as installer>: ")
                else:   
                    fileName=download_repository( logstashRepo )
                unzipGit(fileName)
            
                logstashLocation = input("Enter the Logstash location to store the piepline files in: ")
                update=input_bool("Are you upgrading an existing Corelight Logstsh Pipeline?", default=False)
                if not update:
                    installLogstash(logstashLocation)
                    logstashVersion = input_bool("Are you running Logstash version 8.x or higher?", default=True)
                    raw = input_bool("Do you want to keep the raw message, this will incerease storage space?", default=False)
                    tcp = input_bool("Are you sending data to logstsh over JSON over TCP?:", default=False)
                    if tcp:
                        ssl = input_bool("Will you be enableing SSL?", default=False)
                        if ssl:
                            enableIngest("ssl", raw, logstashLocation)
                        else:
                            enableIngest("tcp", raw, logstashLocation)
                    else:
                        kafka = input_bool("Are sending data to Logstsh via Kafka?:", default=False)
                        if kafka:
                            enableIngest("kafka", raw, logstashLocation)
                        else:
                            hec = input_bool("Are sending data to logstsh over HTTP Event Collector?:", default=False)
                            if hec:
                                enableIngest("hec", raw, logstashLocation)
                else:
                    updateLogstash(logstashLocation)
            else:
                print("Please run script again on Logstash server")
                print("Strange things are afoot at the Circle-K.")
                sys.exit(1)
        else:
            #Answer Yes if you are in a offline or Airgaped enviorment 
            # Then give the filename and location and it will run
            download = input_bool("Have you downloaded Ingest repo?", default=False)
            if download:
                fileName = input("Please enter the filename for the Zip file of the Logstash Pipeline repo? <note it need to be in the same directorey as installer>: ")
            else: 
                fileName=download_repository( ingestRepo )
            unzipGit(fileName)
            uploadIngestPipelines(session,baseURI)
    templateDS = input_bool("Will you be useing Datastreams?", default=True)
    if templateDS:
        datastreams(session,baseURI,logstash,updateTemplates)
        if logstash and not updateLogstash:
            postPorcessing(logstashLocation, templateDS, logstashVersion)
    else:
        templateComponent = input_bool("Will you be useing Component Templates?", default=True)
        if templateComponent:
            component( session, baseURI, logstash, updateTemplates )
            if logstash and not updateLogstash:
                postPorcessing(logstashLocation, templateDS, logstashVersion)
        else:
            templateLegcy = input_bool("Will you be using Legcy Templates? This is not supported on version 8.x and above?", default=False)
            if templateLegcy:
                index(session,baseURI,logstash,updateTemplates)
                if logstash and not updateLogstash:
                    postPorcessing(logstashLocation, templateDS, logstashVersion)

if __name__ == "__main__":
    main()
